System/OsString/Common.hs

```diff
@@ -313,3 +410,779 @@ toChar (WindowsChar w) = chr $ fromIntegral w
+#ifdef WINDOWS
+head (WindowsString s) = WindowsChar (BS16.head s)
+#else
+head (PosixString s) = PosixChar (BS.head s)
+#endif
+
+-- | /O(n)/ Return all the elements of a 'OsString' except the last one.
+-- An exception will be thrown in the case of an empty OsString.
+--
+-- This is a partial function, consider using 'unsnoc' instead.
+--
+-- @since 1.4.200.0
+init :: HasCallStack => PLATFORM_STRING -> PLATFORM_STRING
+#ifdef WINDOWS
+init (WindowsString s) = WindowsString (BS16.init s)
+#else
+init (PosixString s) = PosixString (BS.init s)
+#endif
+
+-- | /O(n)/ Extract the 'init' and 'last' of a OsString, returning 'Nothing'
+-- if it is empty.
+--
+-- @since 1.4.200.0
+unsnoc :: PLATFORM_STRING -> Maybe (PLATFORM_STRING, PLATFORM_WORD)
+#ifdef WINDOWS
+unsnoc (WindowsString s) = (bimap WindowsString WindowsChar) <$> (BS16.unsnoc s)
+#else
+unsnoc (PosixString s) = (bimap PosixString PosixChar) <$> (BS.unsnoc s)
+#endif
+
+-- | /O(1)/. The empty 'OsString'.
+--
+-- @since 1.4.200.0
+null :: PLATFORM_STRING -> Bool
+#ifdef WINDOWS
+null (WindowsString s) = BS16.null s
+#else
+null (PosixString s) = BS.null s
+#endif
+
+-- | /O(1)/ The length of a 'OsString'.
+--
+-- @since 1.4.200.0
+length :: PLATFORM_STRING -> Int
+#ifdef WINDOWS
+length (WindowsString s) = BS16.length s
+#else
+length (PosixString s) = BS.length s
+#endif
+
+-- | /O(n)/ 'map' @f xs@ is the OsString obtained by applying @f@ to each
+-- element of @xs@.
+--
+-- @since 1.4.200.0
+map :: (PLATFORM_WORD -> PLATFORM_WORD) -> PLATFORM_STRING -> PLATFORM_STRING
+#ifdef WINDOWS
+map f (WindowsString s) = WindowsString (BS16.map (getWindowsChar . f . WindowsChar) s)
+#else
+map f (PosixString s) = PosixString (BS.map (getPosixChar . f . PosixChar) s)
+#endif
+
+-- | /O(n)/ 'reverse' @xs@ efficiently returns the elements of @xs@ in reverse order.
+--
+-- @since 1.4.200.0
+reverse :: PLATFORM_STRING -> PLATFORM_STRING
+#ifdef WINDOWS
+reverse (WindowsString s) = WindowsString (BS16.reverse s)
+#else
+reverse (PosixString s) = PosixString (BS.reverse s)
+#endif
+
+-- | /O(n)/ The 'intercalate' function takes a 'OsString' and a list of
+-- 'OsString's and concatenates the list after interspersing the first
+-- argument between each element of the list.
+--
+-- @since 1.4.200.0
+intercalate :: PLATFORM_STRING -> [PLATFORM_STRING] -> PLATFORM_STRING
+#ifdef WINDOWS
+intercalate (WindowsString s) xs = WindowsString (BS16.intercalate s (fmap getWindowsString xs))
+#else
+intercalate (PosixString s) xs = PosixString (BS.intercalate s (fmap getPosixString xs))
+#endif
+
+-- | 'foldl', applied to a binary operator, a starting value (typically
+-- the left-identity of the operator), and a OsString, reduces the
+-- OsString using the binary operator, from left to right.
+--
+-- @since 1.4.200.0
+foldl :: (a -> PLATFORM_WORD -> a) -> a -> PLATFORM_STRING -> a
+#ifdef WINDOWS
+foldl f a (WindowsString s) = BS16.foldl (\a' c -> f a' (WindowsChar c)) a s
```

Comment:
`coerce f` could be better in certain circumstances.

Reply:
Does that also go for 

```hs
unfoldrN n f a = first PosixString $ BS.unfoldrN n (fmap (first getPosixChar) . f) a
```

vs 

```hs
unfoldrN n f a = first PosixString $ BS.unfoldrN n (coerce . f) a
```

Reply:
I'm AFK, but I think pretty much everything in this commit is just `coerce` of a `BS.foo` or `BS16.foo`. It's less error-prone than wrapping-unwrapping through newtypes manually and much better for compiler.

Reply:
Done